src/community/wps-longitudinal-profile/src/main/java/org/geoserver/wps/longitudinal/LongitudinalProfileProcess.java

```diff
@@ -174,27 +170,79 @@ public LongitudinalProfileProcessResult execute(
         Geometry previousPoint = null;
 
         CoordinateReferenceSystem coverageCrs = coverageInfo.getCRS();
-        List<Position2D> positions2D =
-                Arrays.stream(denseLine.getCoordinates())
-                        .map(coordinate -> new Position2D(coverageCrs, coordinate.x, coordinate.y))
-                        .collect(Collectors.toList());
 
-        for (Position2D position2D : positions2D) {
-            LOGGER.fine("processing position:" + position2D);
+        // Create a stack with all geometry vertices
+        Coordinate[] coords = denseLine.getCoordinates();
+        Stack<ProfileVertice> vertices = new Stack<ProfileVertice>();
+        for (int i = 0; i < coords.length; i++) {
+            vertices.add(
+                    new ProfileVertice(
+                            i, new Position2D(coverageCrs, coords[i].x, coords[i].y), null));
+        }
+
+        // Divide vertices stack in chunks to process altitude reading with threads depending of
+        // available cores
+        int cores = Runtime.getRuntime().availableProcessors();
+        int total = vertices.size();
+        int chunkSize = total / cores;
+        int overflow = total % cores;
+
+        Stack<Vector<ProfileVertice>> chunks = new Stack<Vector<ProfileVertice>>();
+        for (int i = 0; i < cores; i++) {
+            Vector<ProfileVertice> chunk = new Vector<ProfileVertice>();
+            for (int j = 0; j < chunkSize; j++) {
+                chunk.add(vertices.pop());
+            }
+            if (overflow > 0) {
+                chunk.add(vertices.pop());
+                overflow = overflow - 1;
+            }
+            chunks.add(chunk);
+        }
+
+        // Create threads executor
+        ExecutorService executor = Executors.newFixedThreadPool(cores);
+        Stack<Future<Vector<ProfileVertice>>> treated = new Stack<Future<Vector<ProfileVertice>>>();
+
+        GridCoverage2DReader gridCoverageReader =
+                (GridCoverage2DReader) coverageInfo.getGridCoverageReader(null, null);
+        GridCoverage2D gridCoverage2D = gridCoverageReader.read(null);
+
+        // Process parallel altitude reading
+        for (int i = 0; i < cores; i++) {
+            treated.add(
+                    executor.submit(
+                            new AltitudeReaderThread(
+                                    chunks.pop(),
+                                    altitudeIndex,
+                                    adjustmentFeatureSource,
+                                    altitudeName,
+                                    gridCoverage2D)));
+        }
+        Vector<ProfileVertice> result = new Vector<ProfileVertice>();
```

Comment:
Usage of Vector is not allowed by GeoServer QA rules. This is a community module, so the PMD checks are not allowed, but it would be best to avoid its usage anyways, if you have any goal to see this module become a supported extension, in the future. A simple ArrayList will do.